In [225]:
import pandas as pd
import numpy as np


In [226]:
data = pd.read_csv('global_freelancers_raw.csv')
data.head()

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
0,FL250001,Ms. Nicole Kidd,f,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN
1,FL250002,Vanessa Garcia,FEMALE,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%
2,FL250003,Juan Nelson,male,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%
3,FL250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%
4,FL250005,Lynn Curtis DDS,female,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%


In [227]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   freelancer_ID        1000 non-null   object 
 1   name                 1000 non-null   object 
 2   gender               1000 non-null   object 
 3   age                  970 non-null    float64
 4   country              1000 non-null   object 
 5   language             1000 non-null   object 
 6   primary_skill        1000 non-null   object 
 7   years_of_experience  949 non-null    float64
 8   hourly_rate (USD)    906 non-null    object 
 9   rating               899 non-null    float64
 10  is_active            911 non-null    object 
 11  client_satisfaction  824 non-null    object 
dtypes: float64(3), object(9)
memory usage: 93.9+ KB


In [228]:
df = data.copy()

In [229]:
df.head()

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
0,FL250001,Ms. Nicole Kidd,f,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN
1,FL250002,Vanessa Garcia,FEMALE,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%
2,FL250003,Juan Nelson,male,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%
3,FL250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%
4,FL250005,Lynn Curtis DDS,female,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%


In [230]:
df.describe()

,age,years_of_experience,rating
count,970.000000,949.000000,899.000000
mean,40.509278,11.340358,2.512570
std,11.942605,9.680610,1.546599
min,20.000000,0.000000,0.000000
25%,31.000000,3.000000,1.400000
50%,41.000000,9.000000,2.600000
75%,51.000000,17.000000,3.800000
max,60.000000,41.000000,5.000000


In [231]:
df.duplicated().sum()

np.int64(0)

In [232]:
df.isna().sum()

freelancer_ID            0
name                     0
gender                   0
age                     30
country                  0
language                 0
primary_skill            0
years_of_experience     51
hourly_rate (USD)       94
rating                 101
is_active               89
client_satisfaction    176
dtype: int64

In [233]:
for col in df:
    print(df[col].unique())

['FL250001' 'FL250002' 'FL250003' 'FL250004' 'FL250005' 'FL250006'
 'FL250007' 'FL250008' 'FL250009' 'FL250010' 'FL250011' 'FL250012'
 'FL250013' 'FL250014' 'FL250015' 'FL250016' 'FL250017' 'FL250018'
 'FL250019' 'FL250020' 'FL250021' 'FL250022' 'FL250023' 'FL250024'
 'FL250025' 'FL250026' 'FL250027' 'FL250028' 'FL250029' 'FL250030'
 'FL250031' 'FL250032' 'FL250033' 'FL250034' 'FL250035' 'FL250036'
 'FL250037' 'FL250038' 'FL250039' 'FL250040' 'FL250041' 'FL250042'
 'FL250043' 'FL250044' 'FL250045' 'FL250046' 'FL250047' 'FL250048'
 'FL250049' 'FL250050' 'FL250051' 'FL250052' 'FL250053' 'FL250054'
 'FL250055' 'FL250056' 'FL250057' 'FL250058' 'FL250059' 'FL250060'
 'FL250061' 'FL250062' 'FL250063' 'FL250064' 'FL250065' 'FL250066'
 'FL250067' 'FL250068' 'FL250069' 'FL250070' 'FL250071' 'FL250072'
 'FL250073' 'FL250074' 'FL250075' 'FL250076' 'FL250077' 'FL250078'
 'FL250079' 'FL250080' 'FL250081' 'FL250082' 'FL250083' 'FL250084'
 'FL250085' 'FL250086' 'FL250087' 'FL250088' 'FL250089' 'FL250

In [234]:
df.columns

Index(['freelancer_ID', 'name', 'gender', 'age', 'country', 'language',
       'primary_skill', 'years_of_experience', 'hourly_rate (USD)', 'rating',
       'is_active', 'client_satisfaction'],
      dtype='object')

Insights gathered
- []  freelancer_ID if want we can remmove the FL Which is in the start and there's no missing values and remove the extra white space and conver it into numeric 
- [] name  make it standardized and remove the extra spaces make the data type to string it has no missing values 
- [] gender we can feature scale this as we can see there many format so keep it short like M and F for all those format and if want you could scale it to 0 and 1 so that it will be easy for the model to understand
- [] age handle the missing values and nothign much in min and max we have seen it has no negative values and any outliers too
- [] country no nul values make it standardized and remove the extra spaces  dtype to string
- [] language same no null value smake it standardized and remove extra spaces dt ype to string
- [] primary_skill  no missing values make it standardized and remove extra spcae and change the stype to string 
- [] years_of_experience it has null values analyze them and follow accordingly else change it to unknown better if ddn't find any relation to other column else we can mediain of the age as the age is raltewd to year of  experiecne 
- []hourly_rate (USD) we have many format where it includes teh dfollar symbols and mention the USD remove all thsoe and make just number and make the dtype to int handle the missing values to unknown or we take median of yeears of experince median like year of experince linkt o hourpay can get the hourly rate too 
- [] rating we could make it median or just fill unknown the missing values
- [] is_active have many format so make 0 to 1 like yes to 1 and no to 0 and we can fill the null by inspwecting like we don't know whether we can fill the midain or keep the naa values as no
- [] client_satisfaction has many missing values we can either get the median or make it null only and remove the percentage sumbol make it to int and mention the %b symbol in colum name

Claude given insights : almost matched with my insights

- [] freelancer_ID — no missing values; remove "FL" prefix if needed, but keep as string/category (not numeric) since it's an identifier, not a value to do math on. Just trim whitespace.
- []  name — no missing values; standardize casing, strip extra whitespace, keep dtype as string.
- []  gender — many formats (f, FEMALE, male, M, etc.); standardize to just M/F, optionally encode to 0/1 for modeling later.
- []  age — has 30 missing values; no negative values or outliers (range 20–60). Fill missing values (median is reasonable here since no outliers).
- []  country — no missing values; standardize casing, remove extra whitespace, dtype string.
- []  language — no missing values; standardize casing, remove extra whitespace, dtype string.
- []  primary_skill — no missing values; standardize casing, remove extra whitespace, dtype string.
- []  years_of_experience — has missing values; fill using median grouped by age bucket (e.g. 20s/30s/40s) since age and experience are related, rather than one global median.
- [] hourly_rate (USD) — mixed formats ($40, USD 100, plain 30); strip $/USD, convert to float (not int, in case of decimals); fill missing values using median grouped by years_of_experience.
- []  rating — has missing values; fill with median, or mark as unknown if you don't want to assume.
- [] is_active — many formats (Yes/No, Y/N, True/False, 1/0); standardize to 0/1; fill missing values with mode (most frequent value), not median, since it's binary.
- [] client_satisfaction — has the most missing values (176); mixed format (84% vs plain 92 with no %); strip % symbol, convert to numeric, fill missing with median or leave as null; rename column to something like client_satisfaction_% to indicate units.

freelancer_ID — no missing values; remove "FL" prefix if needed, but keep as string/category (not numeric) since it's an identifier, not a value to do math on. Just trim whitespace.

In [235]:
df['freelancer_ID'].isna().sum()
df['freelancer_ID']  = df['freelancer_ID'].str.replace('FL' , '')
df


,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
0,250001,Ms. Nicole Kidd,f,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN
1,250002,Vanessa Garcia,FEMALE,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%
2,250003,Juan Nelson,male,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%
3,250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%
4,250005,Lynn Curtis DDS,female,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%
...,...,...,...,...,...,...,...,...,...,...,...,...
995,250996,Albert Wilcox,Male,56.0,Turkey,Turkish,DevOps,13.0,100,0.0,no,68%
996,250997,Cheryl Norris,f,26.0,Germany,German,Blockchain Development,6.0,USD 40,2.8,N,82
997,250998,Kathy Watkins,female,37.0,Japan,Japanese,Data Analysis,15.0,75,NaN,False,94%
998,250999,John Obrien,m,46.0,Russia,Russian,Machine Learning,22.0,100,2.8,yes,97


In [236]:
df['freelancer_ID'].str.contains('FL').sum()


np.int64(0)

In [237]:
df['freelancer_ID'] = df['freelancer_ID'].str.strip()

In [238]:
df['freelancer_ID'].str.contains(r'^\s|s$', regex=True).sum() # ensuring that there are no white spaces

np.int64(0)

name — no missing values; standardize casing, strip extra whitespace, keep dtype as string.

In [239]:
df['name'].isna().sum()

np.int64(0)

In [240]:
df['name'].duplicated().sum()

np.int64(8)

In [241]:
df[df['name'].duplicated(keep= False)].sort_values('name')

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
669,250670,Amanda Mitchell,FEMALE,25.0,Canada,English,DevOps,1.0,USD 30,3.2,0,87%
692,250693,Amanda Mitchell,female,31.0,France,French,Mobile Apps,NaN,50,1.4,no,80%
295,250296,Amy Lee,FEMALE,56.0,United Kingdom,English,DevOps,34.0,NaN,2.1,yes,84%
486,250487,Amy Lee,f,48.0,Russia,Russian,Blockchain Development,13.0,$100,2.7,NaN,62%
125,250126,Jason Carter,Male,49.0,South Korea,Korean,Data Analysis,19.0,NaN,5.0,Y,83%
416,250417,Jason Carter,M,38.0,Mexico,Spanish,Data Analysis,9.0,$50,0.0,0,74%
5,250006,Lisa Johnson,female,59.0,Netherlands,Dutch,AI,14.0,$30,2.4,False,NaN
689,250690,Lisa Johnson,Female,52.0,Canada,English,UI/UX Design,25.0,100,4.8,N,NaN
377,250378,Mary Brown,female,23.0,Germany,German,Machine Learning,4.0,USD 20,4.7,True,84%
694,250695,Mary Brown,FEMALE,53.0,South Africa,Afrikaans,Graphic Design,35.0,NaN,0.0,1,67%


In [242]:
df['name'] = df['name'].str.title()

In [243]:
df['name'] = df['name'].str.strip()

In [244]:
df[df['name'].str.contains(r'^\s|s$')] # this it is for checking the middle spaces which they have.

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
4,250005,Lynn Curtis Dds,female,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%
6,250007,Eric Myers,m,52.0,Indonesia,Indonesian,Data Analysis,10.0,USD 75,3.1,0,94%
11,250012,Michelle Curtis,FEMALE,55.0,Argentina,Spanish,UI/UX Design,15.0,NaN,NaN,Y,86%
17,250018,Scott Burns,M,56.0,India,Hindi,Blockchain Development,NaN,NaN,NaN,NaN,87%
26,250027,Marissa Bass,female,34.0,Japan,Japanese,Blockchain Development,8.0,$20,0.0,True,89%
...,...,...,...,...,...,...,...,...,...,...,...,...
989,250990,Evan Ramos,Male,37.0,Australia,English,Graphic Design,4.0,75,3.2,False,76%
990,250991,Michelle Solis,FEMALE,21.0,Argentina,Spanish,AI,2.0,100,3.7,yes,73
993,250994,Jim Weiss,male,46.0,Russia,Russian,Mobile Apps,8.0,20,2.2,NaN,89
996,250997,Cheryl Norris,f,26.0,Germany,German,Blockchain Development,6.0,USD 40,2.8,N,82


In [245]:
df.loc[df['name'].str.contains(r'[^A-Za-z0-9,.\s]',na=False , regex= True),'name'].sum() # this explains that there is no special character

0

In [246]:
df['name']  = df['name'].astype('string')
df['name'].dtype

string[python]

In [247]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   freelancer_ID        1000 non-null   object 
 1   name                 1000 non-null   string 
 2   gender               1000 non-null   object 
 3   age                  970 non-null    float64
 4   country              1000 non-null   object 
 5   language             1000 non-null   object 
 6   primary_skill        1000 non-null   object 
 7   years_of_experience  949 non-null    float64
 8   hourly_rate (USD)    906 non-null    object 
 9   rating               899 non-null    float64
 10  is_active            911 non-null    object 
 11  client_satisfaction  824 non-null    object 
dtypes: float64(3), object(8), string(1)
memory usage: 93.9+ KB


gender — many formats (f, FEMALE, male, M, etc.); standardize to just M/F, optionally encode to 0/1 for modeling later.

In [248]:
df['gender'].unique()

array(['f', 'FEMALE', 'male', 'F', 'female', 'm', 'MALE', 'Female', 'M',
       'Male'], dtype=object)

In [249]:
df['gender'] = df['gender'].replace({

    'f' : 'F', 'FEMALE' : 'F', 'female': 'F','Female': 'F',

     'male': 'M' , 'm' : 'M' , 'MALE':'M' , 'Male' : 'M'
    })


In [250]:
df['gender'].unique()

array(['F', 'M'], dtype=object)

In [251]:
df.head()

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
0,250001,Ms. Nicole Kidd,F,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN
1,250002,Vanessa Garcia,F,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%
2,250003,Juan Nelson,M,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%
3,250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%
4,250005,Lynn Curtis Dds,F,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%


age — has 30 missing values; no negative values or outliers (range 20–60). Fill missing values (median is reasonable here since no outliers).

In [252]:
df['age'].describe()

count    970.000000
mean      40.509278
std       11.942605
min       20.000000
25%       31.000000
50%       41.000000
75%       51.000000
max       60.000000
Name: age, dtype: float64

In [253]:
df['age'].median()
df[df['age'].isna()]

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
20,250021,James Cherry,M,NaN,Indonesia,Indonesian,AI,9.0,50,1.8,False,65%
33,250034,Michael Anderson,M,NaN,South Korea,Korean,AI,9.0,40,2.4,yes,82
53,250054,Jimmy Hogan,M,NaN,Spain,Spanish,UI/UX Design,1.0,30,0.0,0,68%
68,250069,Corey Warren,M,NaN,Argentina,Spanish,UI/UX Design,6.0,100,1.5,False,NaN
82,250083,Andrew Avila,M,NaN,China,Mandarin,DevOps,3.0,40,4.6,N,74%
98,250099,Andrew Lynch,M,NaN,United States,English,UI/UX Design,5.0,$20,1.1,no,74%
133,250134,Cindy Jacobs,F,NaN,Brazil,Portuguese,Machine Learning,4.0,NaN,2.5,NaN,70%
150,250151,Matthew Richards,M,NaN,Turkey,Turkish,Data Analysis,10.0,USD 40,0.0,NaN,NaN
161,250162,Donald Mitchell,M,NaN,Argentina,Spanish,Data Analysis,12.0,$30,4.1,no,68%
170,250171,Debra Salazar,F,NaN,Netherlands,Dutch,Blockchain Development,NaN,40,2.8,N,80%


In [254]:
df

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction
0,250001,Ms. Nicole Kidd,F,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN
1,250002,Vanessa Garcia,F,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%
2,250003,Juan Nelson,M,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%
3,250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%
4,250005,Lynn Curtis Dds,F,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%
...,...,...,...,...,...,...,...,...,...,...,...,...
995,250996,Albert Wilcox,M,56.0,Turkey,Turkish,DevOps,13.0,100,0.0,no,68%
996,250997,Cheryl Norris,F,26.0,Germany,German,Blockchain Development,6.0,USD 40,2.8,N,82
997,250998,Kathy Watkins,F,37.0,Japan,Japanese,Data Analysis,15.0,75,NaN,False,94%
998,250999,John Obrien,M,46.0,Russia,Russian,Machine Learning,22.0,100,2.8,yes,97


In [255]:
mising_values = df['age'].isna()

In [256]:
df['years_of_experience'].isna().sum()

np.int64(51)

In [257]:
df['exp_group'] = pd.cut(df['years_of_experience'] , bins = [-1,2,5,20,30,100], labels=['Beginner','Junior','Mid','Professional','Veteran'])
df['age'] = df.groupby('exp_group')['age'].transform(lambda x : x.fillna(x.median()))
df['age'] = df['age'].fillna(df['age'].median())

/var/folders/qw/_nk9z1hd74q9vjp80k85r5l80000gn/T/ipykernel_73665/1661787127.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['age'] = df.groupby('exp_group')['age'].transform(lambda x : x.fillna(x.median()))


In [258]:
df[['years_of_experience','exp_group']]

,years_of_experience,exp_group
0,11.0,Mid
1,34.0,Veteran
2,31.0,Veteran
3,4.0,Junior
4,27.0,Professional
...,...,...
995,13.0,Mid
996,6.0,Mid
997,15.0,Mid
998,22.0,Professional


In [259]:
df['age'].isna().sum()

np.int64(0)

In [260]:

df[mising_values][['age','years_of_experience','exp_group']]


,age,years_of_experience,exp_group
20,41.0,9.0,Mid
33,41.0,9.0,Mid
53,26.0,1.0,Beginner
68,41.0,6.0,Mid
82,34.0,3.0,Junior
98,34.0,5.0,Junior
133,34.0,4.0,Junior
150,41.0,10.0,Mid
161,41.0,12.0,Mid
170,41.0,NaN,NaN


In [261]:
df[['age','years_of_experience','exp_group']]

,age,years_of_experience,exp_group
0,52.0,11.0,Mid
1,52.0,34.0,Veteran
2,53.0,31.0,Veteran
3,38.0,4.0,Junior
4,53.0,27.0,Professional
...,...,...,...
995,56.0,13.0,Mid
996,26.0,6.0,Mid
997,37.0,15.0,Mid
998,46.0,22.0,Professional


country — no missing values; standardize casing, remove extra whitespace, dtype string.

In [262]:
df['country'].isna().sum()

np.int64(0)

In [263]:
df['country'] = df['country'].str.title()

In [264]:
df['country'] = df['country'].str.strip()

In [265]:
df[df['country'].str.contains(r'^\s|\s$',regex=True)]

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction,exp_group


In [266]:
df['country'] = df['country'].astype('string')

In [267]:
df['country'].dtype

string[python]

language — no missing values; standardize casing, remove extra whitespace, dtype string.
primary_skill — no missing values; standardize casing, remove extra whitespace, dtype string.

In [268]:
df['language'].isna().sum()

np.int64(0)

In [269]:
df['language'] = df['language'].str.title()

In [270]:
df['language'] = df['language'].str.strip()

In [271]:
df[df['language'].str.contains(r'^\s|\s$',regex=True)]

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction,exp_group


In [272]:
df['primary_skill'].isna().sum()

np.int64(0)

In [273]:
df['primary_skill'] = df['primary_skill'].str.title()

In [274]:
df['primary_skill'] = df['primary_skill'].str.strip()

In [275]:
df

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction,exp_group
0,250001,Ms. Nicole Kidd,F,52.0,Italy,Italian,Blockchain Development,11.0,100,NaN,0,NaN,Mid
1,250002,Vanessa Garcia,F,52.0,Australia,English,Mobile Apps,34.0,USD 100,3.3,1,84%,Veteran
2,250003,Juan Nelson,M,53.0,Germany,German,Graphic Design,31.0,50,0.0,N,71%,Veteran
3,250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,$40,1.5,N,90%,Junior
4,250005,Lynn Curtis Dds,F,53.0,Germany,German,Web Development,27.0,30,4.8,0,83%,Professional
...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,250996,Albert Wilcox,M,56.0,Turkey,Turkish,Devops,13.0,100,0.0,no,68%,Mid
996,250997,Cheryl Norris,F,26.0,Germany,German,Blockchain Development,6.0,USD 40,2.8,N,82,Mid
997,250998,Kathy Watkins,F,37.0,Japan,Japanese,Data Analysis,15.0,75,NaN,False,94%,Mid
998,250999,John Obrien,M,46.0,Russia,Russian,Machine Learning,22.0,100,2.8,yes,97,Professional


years_of_experience — has missing values; fill using median grouped by age bucket (e.g. 20s/30s/40s) since age and experience are related, rather than one global median.

In [276]:
missing_values = df['years_of_experience'].isna()
missing_values.sum()

np.int64(51)

In [277]:
df['age_group'] = pd.cut(df['age'], bins=[19,29,39,49,60], labels=['20s','30s','40s','50s'])
grouped_median = df.groupby('age_group')['years_of_experience'].transform(lambda x : x.fillna(x.median()))
df['years_of_experience'] =df['years_of_experience'].fillna(grouped_median)


/var/folders/qw/_nk9z1hd74q9vjp80k85r5l80000gn/T/ipykernel_73665/3711717219.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_median = df.groupby('age_group')['years_of_experience'].transform(lambda x : x.fillna(x.median()))


In [278]:
df['age_group'].isna().sum()

np.int64(0)

In [279]:
df['years_of_experience'].isna().sum()

np.int64(0)

In [280]:
df[missing_values][['age','years_of_experience','age_group']]

,age,years_of_experience,age_group
17,41.0,12.0,40s
21,41.0,12.0,40s
43,41.0,12.0,40s
58,41.0,12.0,40s
75,41.0,12.0,40s
157,41.0,12.0,40s
165,41.0,12.0,40s
170,41.0,12.0,40s
189,41.0,12.0,40s
192,41.0,12.0,40s


In [281]:
print(df['years_of_experience'].unique())

[11. 34. 31.  4. 27. 14. 10. 22. 17. 15. 18. 20.  0. 13. 12.  3.  9. 21.
  8. 19. 32. 28.  6. 23.  7.  1. 33. 16. 35. 38.  5. 30.  2. 25. 39. 29.
 40. 26. 24. 37. 41.]


hourly_rate (USD) — mixed formats ($40, USD 100, plain 30); strip $/USD, convert to float (not int, in case of decimals); fill missing values using median grouped by years_of_experience.

In [282]:
df['hourly_rate (USD)']

0          100
1      USD 100
2           50
3          $40
4           30
        ...   
995        100
996     USD 40
997         75
998        100
999        $20
Name: hourly_rate (USD), Length: 1000, dtype: object

In [283]:
df['hourly_rate (USD)'].unique()

array(['100', 'USD 100', '50', '$40', '30', '$30', 'USD 75', 'USD 40',
       nan, '$50', '40', '75', 'USD 50', 'USD 30', '$20', '20', '$75',
       '$100', 'USD 20'], dtype=object)

In [284]:
df['hourly_rate (USD)'] = df['hourly_rate (USD)'].str.replace(r'USD|\$' , '' , regex= True)

In [285]:
df['hourly_rate (USD)'].unique()

array(['100', ' 100', '50', '40', '30', ' 75', ' 40', nan, '75', ' 50',
       ' 30', '20', ' 20'], dtype=object)

In [286]:
df['hourly_rate (USD)']
missing_hours=df['hourly_rate (USD)'].isna()

In [287]:
df[df['hourly_rate (USD)'].isna()]
df['hourly_rate (USD)'] = df['hourly_rate (USD)'].astype('float')
df['hourly_rate (USD)'].dtype

dtype('float64')

In [288]:
grouped_median = df.groupby('exp_group')['hourly_rate (USD)'].transform(lambda x : x.fillna(x.median()))
df['hourly_rate (USD)'] =df['hourly_rate (USD)'].fillna(grouped_median)
df['hourly_rate (USD)'] =df['hourly_rate (USD)'].fillna(df['hourly_rate (USD)'].median())


/var/folders/qw/_nk9z1hd74q9vjp80k85r5l80000gn/T/ipykernel_73665/1999403151.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  grouped_median = df.groupby('exp_group')['hourly_rate (USD)'].transform(lambda x : x.fillna(x.median()))


In [289]:
df['hourly_rate (USD)'].isna().sum()

np.int64(0)

In [290]:
df[missing_hours][['age','hourly_rate (USD)' , 'years_of_experience' , 'exp_group']]

,age,hourly_rate (USD),years_of_experience,exp_group
8,26.0,40.0,4.0,Junior
11,55.0,40.0,15.0,Mid
12,57.0,40.0,18.0,Mid
17,41.0,40.0,12.0,NaN
21,41.0,40.0,12.0,NaN
...,...,...,...,...
934,38.0,40.0,11.0,Mid
937,21.0,40.0,0.0,Beginner
945,24.0,40.0,2.0,Beginner
969,49.0,40.0,1.0,Beginner


In [291]:
df['exp_group'] = pd.cut(df['years_of_experience'], bins=[-1,2,15,20,25,100], labels=['Beginner','Junior','Mid','Professional','Veteran'])
print(df['exp_group'].isnull().sum())  # should now be 0

0


In [292]:
df

,freelancer_ID,name,gender,age,country,language,primary_skill,years_of_experience,hourly_rate (USD),rating,is_active,client_satisfaction,exp_group,age_group
0,250001,Ms. Nicole Kidd,F,52.0,Italy,Italian,Blockchain Development,11.0,100.0,NaN,0,NaN,Junior,50s
1,250002,Vanessa Garcia,F,52.0,Australia,English,Mobile Apps,34.0,100.0,3.3,1,84%,Veteran,50s
2,250003,Juan Nelson,M,53.0,Germany,German,Graphic Design,31.0,50.0,0.0,N,71%,Veteran,50s
3,250004,Amanda Spencer,F,38.0,Australia,English,Web Development,4.0,40.0,1.5,N,90%,Junior,30s
4,250005,Lynn Curtis Dds,F,53.0,Germany,German,Web Development,27.0,30.0,4.8,0,83%,Veteran,50s
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,250996,Albert Wilcox,M,56.0,Turkey,Turkish,Devops,13.0,100.0,0.0,no,68%,Junior,50s
996,250997,Cheryl Norris,F,26.0,Germany,German,Blockchain Development,6.0,40.0,2.8,N,82,Junior,20s
997,250998,Kathy Watkins,F,37.0,Japan,Japanese,Data Analysis,15.0,75.0,NaN,False,94%,Junior,30s
998,250999,John Obrien,M,46.0,Russia,Russian,Machine Learning,22.0,100.0,2.8,yes,97,Professional,40s


In [293]:
df['hourly_rate (USD)'].isna().sum()

np.int64(0)

rating — has missing values; fill with median, or mark as unknown if you don't want to assume.

In [294]:
df['rating'].isna().sum()

np.int64(101)

In [295]:
df['rating'] = df['rating'].fillna(df['rating'].median())

In [296]:
df['rating'].isna().sum()

np.int64(0)

is_active — many formats (Yes/No, Y/N, True/False, 1/0); standardize to 0/1; fill missing values with mode (most frequent value), not median, since it's binary.

In [297]:
df['is_active'].unique()

array(['0', '1', 'N', 'False', 'True', 'yes', 'Y', nan, 'no'],
      dtype=object)

In [298]:
df['is_active'] = df['is_active'].replace({
    'N' : '0' ,'False' : '0' , 'no' : '0',
    'True' : '1', 'yes' : '1', 'Y' : '1'
})

In [299]:
df['is_active'].unique()

array(['0', '1', nan], dtype=object)

In [300]:
df['is_active'].isna().sum()

np.int64(89)

In [301]:
df['is_active'] = df['is_active'].fillna(df['is_active'].mode()[0])

In [302]:
df['is_active'].isna().sum()

np.int64(0)

In [303]:
df['is_active'].mode()[0]

'0'

In [305]:
df['is_active'].unique()

array(['0', '1'], dtype=object)

client_satisfaction — has the most missing values (176); mixed format (84% vs plain 92 with no %); strip % symbol, convert to numeric, fill missing with median or leave as null; rename column to something like client_satisfaction_% to indicate units.

In [ ]:
df['client_satisfaction'].isna().sum()

np.int64(176)

In [309]:
df['client_satisfaction'] =  df['client_satisfaction'].str.replace('%' , '')

In [310]:
df['client_satisfaction'].unique()

array([nan, '84', '71', '90', '83', '94', '76', '77', '86', '93', '70',
       '69', '60', '87', '75', '68', '65', '100', '92', '89', '62', '82',
       '81', '63', '67', '80', '74', '85', '79', '72', '64', '88', '96',
       '61', '97', '73', '78', '95', '66', '91', '99', '98'], dtype=object)

In [316]:
df['client_satisfaction'] = pd.to_numeric(df['client_satisfaction'],errors='coerce')

df['client_satisfaction']  = df['client_satisfaction'].fillna(df['client_satisfaction'].median())

In [318]:
df['client_satisfaction'].isna().sum()

np.int64(0)

In [319]:
df.describe()

,age,years_of_experience,hourly_rate (USD),rating,client_satisfaction
count,1000.000000,1000.000000,1000.000000,1000.000000,1000.000000
mean,40.482000,11.374000,51.420000,2.521400,79.222000
std,11.565528,9.431388,26.228258,1.466572,10.434059
min,20.000000,0.000000,20.000000,0.000000,60.000000
25%,31.000000,3.000000,30.000000,1.500000,71.000000
50%,41.000000,10.000000,40.000000,2.600000,79.000000
75%,50.000000,17.000000,75.000000,3.700000,87.000000
max,60.000000,41.000000,100.000000,5.000000,100.000000


In [320]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   freelancer_ID        1000 non-null   object  
 1   name                 1000 non-null   string  
 2   gender               1000 non-null   object  
 3   age                  1000 non-null   float64 
 4   country              1000 non-null   string  
 5   language             1000 non-null   object  
 6   primary_skill        1000 non-null   object  
 7   years_of_experience  1000 non-null   float64 
 8   hourly_rate (USD)    1000 non-null   float64 
 9   rating               1000 non-null   float64 
 10  is_active            1000 non-null   object  
 11  client_satisfaction  1000 non-null   float64 
 12  exp_group            1000 non-null   category
 13  age_group            1000 non-null   category
dtypes: category(2), float64(5), object(5), string(2)
memory usage: 96.2+ KB


In [322]:
df['is_active'] = pd.to_numeric(df['is_active'])

In [323]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   freelancer_ID        1000 non-null   object  
 1   name                 1000 non-null   string  
 2   gender               1000 non-null   object  
 3   age                  1000 non-null   float64 
 4   country              1000 non-null   string  
 5   language             1000 non-null   object  
 6   primary_skill        1000 non-null   object  
 7   years_of_experience  1000 non-null   float64 
 8   hourly_rate (USD)    1000 non-null   float64 
 9   rating               1000 non-null   float64 
 10  is_active            1000 non-null   int64   
 11  client_satisfaction  1000 non-null   float64 
 12  exp_group            1000 non-null   category
 13  age_group            1000 non-null   category
dtypes: category(2), float64(5), int64(1), object(4), string(2)
memory usage: 

In [324]:
df.to_csv("cleaned_freelance.csv")